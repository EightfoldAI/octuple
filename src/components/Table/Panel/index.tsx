import React from 'react';
import { TitleProps } from './Panel.types';

function Panel({ classNames, children }: TitleProps) {
    return <div className={classNames}>{children}</div>;
}

export default Panel;
