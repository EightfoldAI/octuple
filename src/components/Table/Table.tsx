import React, {
    forwardRef,
    ReactNode,
    useCallback,
    useContext,
    useMemo,
    useRef,
} from 'react';
import type { Breakpoint } from '../../shared/utilities';
import { mergeClasses, omit, scrollTo } from '../../shared/utilities';
import OcTable, { Summary } from './Internal';
import type { OcTableProps } from './Internal/OcTable.types';
import { convertChildrenToColumns } from './Internal/Hooks/useColumns';
import { Spinner, SpinnerSize } from '../Spinner';
import { Pagination } from '../Pagination';
import usePagination, {
    DEFAULT_PAGE_SIZE,
    getPaginationParam,
} from './Hooks/usePagination';
import useLazyKVMap from './Hooks/useLazyKVMap';
import type {
    ChangeEventInfo,
    ColumnGroupType,
    ColumnType,
    ExpandableConfig,
    ExpandType,
    FilterValue,
    GetRowKey,
    SorterResult,
    TableAction,
    TableProps,
} from './Table.types';
import {
    EMPTY_LIST,
    ColumnsType,
    TablePaginationConfig,
    TableSize,
} from './Table.types';
import useSelection, {
    SELECTION_ALL,
    SELECTION_COLUMN,
    SELECTION_INVERT,
    SELECTION_NONE,
} from './Hooks/useSelection';
import type { SortState } from './Hooks/useSorter';
import useSorter, { getSortData } from './Hooks/useSorter';
import type { FilterState } from './Hooks/useFilter';
import useFilter, { getFilterData } from './Hooks/useFilter';
import useTitleColumns from './Hooks/useTitleColumns';
import renderExpandIcon from './ExpandIcon';
import { Size, SizeContext } from '../ConfigProvider';
import Column from './Internal/Column';
import ColumnGroup from './Internal/ColumnGroup';
import { useBreakpoint } from '../../hooks/useBreakpoint';
import { useCanvasDirection } from '../../hooks/useCanvasDirection';
import { Empty, EmptyMode } from '../Empty/index';

import styles from './Styles/table.module.scss';

export { ColumnsType, TablePaginationConfig };

function InternalTable<RecordType extends object = any>(
    props: TableProps<RecordType>,
    ref: React.MutableRefObject<HTMLDivElement>
) {
    const {
        alternateRowColor = true,
        bordered,
        cancelSortText,
        cellBordered = false,
        children,
        classNames,
        collapseText,
        columns,
        dataSource,
        emptyText,
        emptyTextDetails,
        expandableConfig,
        expandText,
        filterCheckallText,
        filterConfirmText,
        filterEmptyText,
        filterResetText,
        filterSearchPlaceholderText,
        getPopupContainer,
        headerBordered = false,
        headerBottomBordered = false,
        indentSize,
        innerBordered = false,
        loading,
        onChange,
        outerBordered = false,
        pagination,
        rowBordered = false,
        rowClassName,
        rowKey,
        rowSelection,
        scroll,
        selectInvertText,
        selectionAllText,
        selectNoneText,
        showSorterTooltip = true,
        size = TableSize.Medium,
        sortDirections,
        style,
        triggerAscText,
        triggerDescText,
    } = props;

    const baseColumns: ColumnsType<RecordType> = useMemo(
        () =>
            columns ||
            (convertChildrenToColumns(children) as ColumnsType<RecordType>),
        [columns, children]
    );
    const needResponsive: boolean = useMemo(
        () => baseColumns.some((col: ColumnType<RecordType>) => col.responsive),
        [baseColumns]
    );

    const screens: Partial<Record<Breakpoint, boolean>> =
        useBreakpoint(needResponsive);

    const mergedColumns: (
        | ColumnGroupType<RecordType>
        | ColumnType<RecordType>
    )[] = useMemo(() => {
        const matched = new Set(
            Object.keys(screens).filter((m: Breakpoint) => screens[m])
        );

        return baseColumns.filter(
            (c) =>
                !c.responsive ||
                c.responsive.some((r: Breakpoint) => matched.has(r))
        );
    }, [baseColumns, screens]);

    const tableProps = omit(props, [
        'classNames',
        'style',
        'columns',
    ]) as TableProps<RecordType>;

    const contextuallySized: Size = useContext(SizeContext);
    const mergedSize = contextuallySized || size;

    const htmlDir: string = useCanvasDirection();
    const rawData: readonly RecordType[] = dataSource || EMPTY_LIST;

    const mergedExpandableConfig: ExpandableConfig<RecordType> = {
        ...expandableConfig,
    };

    const { childrenColumnName = 'children' } = mergedExpandableConfig;

    const expandType: ExpandType = useMemo<ExpandType>(() => {
        if (rawData.some((item) => (item as any)?.[childrenColumnName])) {
            return 'nest';
        }

        if (expandableConfig?.expandedRowRender) {
            return 'row';
        }

        return null;
    }, [rawData]);

    const internalRefs = {
        body: useRef<HTMLDivElement>(),
    };

    // ============================ RowKey ============================
    const getRowKey = useMemo<GetRowKey<RecordType>>(() => {
        if (typeof rowKey === 'function') {
            return rowKey;
        }

        return (record: RecordType) => (record as any)?.[rowKey as string];
    }, [rowKey]);

    const [getRecordByKey] = useLazyKVMap(
        rawData,
        childrenColumnName,
        getRowKey
    );

    // ============================ Events =============================
    const changeEventInfo: Partial<ChangeEventInfo<RecordType>> = {};

    const triggerOnChange = (
        info: Partial<ChangeEventInfo<RecordType>>,
        action: TableAction,
        reset: boolean = false
    ) => {
        const changeInfo = {
            ...changeEventInfo,
            ...info,
        };

        if (reset) {
            changeEventInfo.resetPagination!();

            // Reset event param
            if (changeInfo.pagination!.currentPage) {
                changeInfo.pagination!.currentPage = 1;
            }

            // Trigger pagination events
            if (pagination) {
                pagination.onCurrentChange?.(
                    changeInfo.pagination!.currentPage!
                );

                const pages: number[] = changeInfo.pagination!.pageSizes!;

                if (pages) {
                    console.log('triggerOnChange pages: ' + pages);
                    for (let i: number = 0; i < pages.length; ++i) {
                        pagination.onSizeChange?.(
                            changeInfo.pagination!.pageSizes[i]
                        );
                    }
                } else {
                    pagination.onSizeChange?.(changeInfo.pagination!.pageSize!);
                }
            }
        }

        if (
            scroll &&
            scroll.scrollToFirstRowOnChange !== false &&
            internalRefs.body.current
        ) {
            scrollTo(0, {
                getContainer: () => internalRefs.body.current!,
            });
        }

        onChange?.(
            changeInfo.pagination!,
            changeInfo.filters!,
            changeInfo.sorter!,
            {
                currentDataSource: getFilterData(
                    getSortData(
                        rawData,
                        changeInfo.sorterStates!,
                        childrenColumnName
                    ),
                    changeInfo.filterStates!
                ),
                action,
            }
        );
    };

    // ============================ Sorter =============================
    const onSorterChange = (
        sorter: SorterResult<RecordType> | SorterResult<RecordType>[],
        sorterStates: SortState<RecordType>[]
    ) => {
        triggerOnChange(
            {
                sorter,
                sorterStates,
            },
            'sort',
            false
        );
    };
    const [transformSorterColumns, sortStates, sorterTitleProps, getSorters] =
        useSorter<RecordType>({
            cancelSortText,
            mergedColumns,
            onSorterChange,
            sortDirections: sortDirections || ['ascend', 'descend'],
            triggerAscText,
            triggerDescText,
            showSorterTooltip,
        });
    const sortedData = useMemo(
        () => getSortData(rawData, sortStates, childrenColumnName),
        [rawData, sortStates]
    );

    changeEventInfo.sorter = getSorters();
    changeEventInfo.sorterStates = sortStates;

    // ============================ Filter ============================
    const onFilterChange = (
        filters: Record<string, FilterValue>,
        filterStates: FilterState<RecordType>[]
    ) => {
        triggerOnChange(
            {
                filters,
                filterStates,
            },
            'filter',
            true
        );
    };

    const [transformFilterColumns, filterStates, getFilters] =
        useFilter<RecordType>({
            mergedColumns,
            filterConfirmText,
            filterResetText,
            filterEmptyText,
            filterCheckallText,
            filterSearchPlaceholderText,
            onFilterChange,
            getPopupContainer,
        });
    const mergedData = getFilterData(sortedData, filterStates);

    changeEventInfo.filters = getFilters();
    changeEventInfo.filterStates = filterStates;

    // ============================ Column ============================
    const columnTitleProps = useMemo(
        () => ({
            ...sorterTitleProps,
        }),
        [sorterTitleProps]
    );
    const [transformTitleColumns] = useTitleColumns(columnTitleProps);

    // ========================== Pagination ==========================
    const onPaginationChange = (
        currentPage: number,
        pageSize: number,
        pageSizes: number[]
    ) => {
        console.log(
            'onPaginationChange pageSize: ' +
                pageSize +
                ' ' +
                'onPaginationChange pageSizes: ' +
                pageSizes
        );
        triggerOnChange(
            {
                pagination: {
                    ...changeEventInfo.pagination,
                    currentPage,
                    pageSize,
                    pageSizes,
                },
            },
            'paginate'
        );
    };

    const [mergedPagination, resetPagination] = usePagination(
        mergedData.length,
        pagination,
        onPaginationChange
    );

    changeEventInfo.pagination =
        pagination === false
            ? {}
            : getPaginationParam(pagination, mergedPagination);

    changeEventInfo.resetPagination = resetPagination;

    // ============================= Data =============================
    const pageData = useMemo<RecordType[]>(() => {
        if (
            pagination === false ||
            (!mergedPagination.pageSize && !mergedPagination.pageSizes)
        ) {
            return mergedData;
        }

        const {
            currentPage = 1,
            total,
            pageSize = DEFAULT_PAGE_SIZE, // Need to figure out how to update this value
            pageSizes,
        } = mergedPagination;

        // Dynamic table data
        if (pageSizes) {
            console.log('Dynamic table data pageSizes: ' + pageSizes);
            for (let i: number = 0; i < pageSizes.length; ++i) {
                if (pageSize === pageSizes[i]) {
                    console.log(
                        'Dynamic table data pageSizes[' +
                            i +
                            '] value: ' +
                            pageSizes[i]
                    );
                    if (mergedData.length < total!) {
                        if (mergedData.length > pageSizes[i]) {
                            return mergedData.slice(
                                (currentPage - 1) * pageSizes[i],
                                currentPage * pageSizes[i]
                            );
                        }
                        return mergedData;
                    }

                    console.log(
                        mergedData.slice(
                            (currentPage - 1) * pageSizes[i],
                            currentPage * pageSizes[i]
                        )
                    );

                    return mergedData.slice(
                        (currentPage - 1) * pageSizes[i],
                        currentPage * pageSizes[i]
                    );
                }
            }
        } else {
            console.log('Dynamic table data pageSize: ' + pageSize);
            if (mergedData.length < total!) {
                if (mergedData.length > pageSize) {
                    return mergedData.slice(
                        (currentPage - 1) * pageSize,
                        currentPage * pageSize
                    );
                }
                return mergedData;
            }

            return mergedData.slice(
                (currentPage - 1) * pageSize,
                currentPage * pageSize
            );
        }
        return null;
    }, [
        !!pagination,
        mergedData,
        mergedPagination?.currentPage,
        mergedPagination?.pageSize,
        mergedPagination?.pageSizes,
        mergedPagination?.total,
    ]);

    // ========================== Selections ==========================
    const [transformSelectionColumns, selectedKeySet] =
        useSelection<RecordType>(rowSelection, {
            data: mergedData,
            emptyText,
            emptyTextDetails,
            pageData,
            getRowKey,
            getRecordByKey,
            expandType,
            childrenColumnName,
            getPopupContainer,
            selectionAllText,
            selectInvertText,
            selectNoneText,
        });

    const internalRowClassName = (
        record: RecordType,
        index: number,
        indent: number
    ) => {
        let mergedRowClassName;
        if (typeof rowClassName === 'function') {
            mergedRowClassName = mergeClasses([
                rowClassName(record, index, indent),
            ]);
        } else {
            mergedRowClassName = mergeClasses([rowClassName]);
        }

        return mergeClasses(
            {
                [styles.tableRowSelected]: selectedKeySet.has(
                    getRowKey(record, index)
                ),
            },
            mergedRowClassName
        );
    };

    // ========================== Expandable ==========================

    mergedExpandableConfig.expandIcon = renderExpandIcon(
        expandText,
        collapseText
    );

    // Indent size
    if (typeof mergedExpandableConfig.indentSize !== 'number') {
        mergedExpandableConfig.indentSize =
            typeof indentSize === 'number' ? indentSize : 15;
    }

    // ============================ Render ============================
    const transformColumns = useCallback(
        (innerColumns: ColumnsType<RecordType>): ColumnsType<RecordType> =>
            transformTitleColumns(
                transformSelectionColumns(
                    transformFilterColumns(transformSorterColumns(innerColumns))
                )
            ),
        [
            transformSorterColumns,
            transformFilterColumns,
            transformSelectionColumns,
        ]
    );

    let topPaginationNode: React.ReactNode;
    let bottomPaginationNode: React.ReactNode;

    if (pagination !== false && mergedPagination?.total) {
        let paginationSize: TablePaginationConfig['pageSize'];
        let paginationSizes: TablePaginationConfig['pageSizes'];

        if (mergedPagination.pageSize) {
            console.log('Table pageSize: ' + mergedPagination.pageSize);
            paginationSize = mergedPagination.pageSize;
        } else {
            paginationSize = undefined;
        }
        if (mergedPagination.pageSizes) {
            console.log('Table pageSizes: ' + mergedPagination.pageSizes);
            paginationSizes = mergedPagination.pageSizes;
        } else {
            paginationSizes = undefined;
        }

        const renderPagination = (position: string) => (
            <Pagination
                {...mergedPagination}
                classNames={mergeClasses([
                    styles.tablePagination,
                    {
                        [styles.tablePaginationLeft]:
                            position.indexOf('left') !== -1,
                    },
                    {
                        [styles.tablePaginationCenter]:
                            position.indexOf('center') !== -1,
                    },
                    {
                        [styles.tablePaginationRight]:
                            position.indexOf('right') !== -1,
                    },
                    mergedPagination.className,
                ])}
                pageSize={paginationSize}
                pageSizes={paginationSizes}
                total={mergedPagination?.total}
            />
        );
        const defaultPosition = htmlDir === 'rtl' ? 'left' : 'right';
        const { position } = mergedPagination;
        if (position !== null && Array.isArray(position)) {
            const topPos = position.find((p) => p.indexOf('top') !== -1);
            const bottomPos = position.find((p) => p.indexOf('bottom') !== -1);
            const isDisable = position.every((p) => `${p}` === 'none');
            if (!topPos && !bottomPos && !isDisable) {
                bottomPaginationNode = renderPagination(defaultPosition);
            }
            if (topPos) {
                topPaginationNode = renderPagination(
                    topPos!.toLowerCase().replace('top', '')
                );
            }
            if (bottomPos) {
                bottomPaginationNode = renderPagination(
                    bottomPos!.toLowerCase().replace('bottom', '')
                );
            }
        } else {
            bottomPaginationNode = renderPagination(defaultPosition);
        }
    }

    if (typeof loading === 'boolean') {
        return <Spinner size={SpinnerSize.Large} />;
    } else if (typeof loading === 'object') {
        return <Spinner size={SpinnerSize.Large} />;
    }

    const renderEmpty = (
        emptyText: ReactNode,
        emptyTextDetails?: string
    ): ReactNode => {
        if (typeof emptyText === 'string') {
            if (emptyTextDetails) {
                return (
                    <Empty
                        title={emptyText}
                        description={emptyTextDetails}
                        mode={EmptyMode.data}
                    />
                );
            }
            return <Empty title={emptyText} mode={EmptyMode.data} />;
        } else {
            return emptyText;
        }
    };

    const wrapperClassNames: string = mergeClasses([
        styles.tableWrapper,
        { [styles.tableWrapperRtl]: htmlDir === 'rtl' },
        classNames,
    ]);
    return (
        <div ref={ref} className={wrapperClassNames} style={style}>
            {topPaginationNode}
            <OcTable<RecordType>
                {...tableProps}
                columns={mergedColumns as OcTableProps<RecordType>['columns']}
                direction={htmlDir}
                expandableConfig={mergedExpandableConfig}
                classNames={mergeClasses([
                    styles.table,
                    { [styles.tableRtl]: htmlDir === 'rtl' },
                    { [styles.tableLarge]: mergedSize === TableSize.Large },
                    { [styles.tableMedium]: mergedSize === TableSize.Medium },
                    { [styles.tableSmall]: mergedSize === TableSize.Small },
                    { [styles.tableAlternate]: alternateRowColor },
                    { [styles.tableBordered]: bordered },
                    {
                        [styles.tableCellBordered]:
                            !bordered &&
                            !rowBordered &&
                            !innerBordered &&
                            cellBordered,
                    },
                    {
                        [styles.tableHeaderBordered]:
                            !bordered &&
                            !rowBordered &&
                            !innerBordered &&
                            headerBordered,
                    },
                    {
                        [styles.tableHeaderBottomBordered]:
                            !bordered &&
                            !rowBordered &&
                            !innerBordered &&
                            !headerBordered &&
                            headerBottomBordered,
                    },
                    {
                        [styles.tableInnerBordered]:
                            !bordered && !rowBordered && innerBordered,
                    },
                    { [styles.tableOuterBordered]: !bordered && outerBordered },
                    {
                        [styles.tableRowBordered]:
                            !bordered && !innerBordered && rowBordered,
                    },
                    { [styles.tableEmpty]: rawData.length === 0 },
                ])}
                data={pageData}
                rowKey={getRowKey}
                rowClassName={internalRowClassName}
                emptyText={renderEmpty(emptyText, emptyTextDetails)}
                transformColumns={
                    transformColumns as OcTableProps<RecordType>['transformColumns']
                }
            />
            {bottomPaginationNode}
        </div>
    );
}

const ForwardTable = forwardRef(InternalTable) as <
    RecordType extends object = any
>(
    props: React.PropsWithChildren<TableProps<RecordType>> & {
        ref?: React.Ref<HTMLDivElement>;
    }
) => React.ReactElement;

type InternalTableType = typeof ForwardTable;

interface TableInterface extends InternalTableType {
    defaultProps?: Partial<TableProps<any>>;
    SELECTION_COLUMN: typeof SELECTION_COLUMN;
    EXPAND_COLUMN: typeof OcTable.EXPAND_COLUMN;
    SELECTION_ALL: 'SELECT_ALL';
    SELECTION_INVERT: 'SELECT_INVERT';
    SELECTION_NONE: 'SELECT_NONE';
    Column: typeof Column;
    ColumnGroup: typeof ColumnGroup;
    Summary: typeof Summary;
}

const Table = ForwardTable as TableInterface;

Table.defaultProps = {
    filterConfirmText: 'OK',
    filterResetText: 'Reset',
    filterEmptyText: 'No filters',
    filterCheckallText: 'Select all items',
    filterSearchPlaceholderText: 'Search in filters',
    emptyText: 'Short Message Here',
    emptyTextDetails:
        'More detail on how might the user be able to get around this',
    selectInvertText: 'Invert current page',
    selectNoneText: 'Clear all data',
    selectionAllText: 'Select all data',
    expandText: 'Expand row',
    collapseText: 'Collapse row',
    triggerDescText: 'Click to sort descending',
    triggerAscText: 'Click to sort ascending',
    cancelSortText: 'Click to cancel sorting',
    rowKey: 'key',
};

Table.SELECTION_COLUMN = SELECTION_COLUMN;
Table.EXPAND_COLUMN = OcTable.EXPAND_COLUMN;
Table.SELECTION_ALL = SELECTION_ALL;
Table.SELECTION_INVERT = SELECTION_INVERT;
Table.SELECTION_NONE = SELECTION_NONE;
Table.Column = Column;
Table.ColumnGroup = ColumnGroup;
Table.Summary = Summary;

export default Table;
