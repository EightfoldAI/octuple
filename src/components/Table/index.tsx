import Table, { TablePaginationConfig } from './Table';

export { ColumnProps } from './Internal/Column';
export {
    ColumnsType,
    ColumnType,
    ColumnGroupType,
    ExpandableConfig,
    FilterConfirmProps,
    FilterValue,
    SorterResult,
    TableProps,
    TableRowSelection,
    TableSize,
} from './Table.types';
export { TablePaginationConfig };

export default Table;
