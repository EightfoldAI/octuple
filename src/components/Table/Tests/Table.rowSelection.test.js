import React from 'react';
import Enzyme, { mount } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import Table from '../index';
import { CheckBox } from '../../Selectors';

Enzyme.configure({ adapter: new Adapter() });

describe('Table.rowSelection', () => {
    window.requestAnimationFrame = (callback) =>
        window.setTimeout(callback, 16);
    window.cancelAnimationFrame = window.clearTimeout;

    const errorSpy = jest.spyOn(console, 'error').mockImplementation(() => {});

    afterEach(() => {
        errorSpy.mockReset();
    });

    afterAll(() => {
        errorSpy.mockRestore();
    });

    beforeAll(() => {
        Object.defineProperty(window, 'matchMedia', {
            writable: true,
            value: jest.fn().mockImplementation((query) => ({
                matches: false,
                media: query,
                onchange: null,
                addListener: jest.fn(), // Deprecated
                removeListener: jest.fn(), // Deprecated
                addEventListener: jest.fn(),
                removeEventListener: jest.fn(),
                dispatchEvent: jest.fn(),
            })),
        });
    });

    const columns = [
        {
            title: 'Name',
            dataIndex: 'name',
        },
    ];

    const data = [
        { key: 0, name: 'Jack' },
        { key: 1, name: 'Lucy' },
        { key: 2, name: 'Tom' },
        { key: 3, name: 'Jerry' },
    ];

    function createTable(props = {}) {
        return (
            <Table
                columns={columns}
                dataSource={data}
                rowSelection={{}}
                {...props}
            />
        );
    }

    function renderedNames(wrapper) {
        return wrapper.find('BodyRow').map((row) => row.props().record.name);
    }

    function getSelections(wrapper) {
        return wrapper
            .find('BodyRow')
            .map((row) => {
                const { key } = row.props().record;
                if (!row.find('input').at(0).props().checked) {
                    return null;
                }

                return key;
            })
            .filter((key) => key !== null);
    }

    function getIndeterminateSelection(wrapper) {
        return wrapper
            .find('BodyRow')
            .map((row) => {
                const { key } = row.props().record;
                return key;
            })
            .filter((key) => key !== null);
    }

    it('select default row', () => {
        const wrapper = mount(
            createTable({ rowSelection: { defaultSelectedRowKeys: [0] } })
        );
        const checkboxes = wrapper.find('input');

        expect(getSelections(wrapper)).toEqual([0]);

        checkboxes.at(1).simulate('change', { target: { checked: false } });
        expect(getSelections(wrapper)).toEqual([]);

        checkboxes.at(0).simulate('change', { target: { checked: true } });
        expect(getSelections(wrapper)).toEqual([0, 1, 2, 3]);

        checkboxes.at(0).simulate('change', { target: { checked: false } });
        expect(getSelections(wrapper)).toEqual([]);
    });

    it('select by checkbox', () => {
        const wrapper = mount(createTable());
        const checkboxes = wrapper.find('input');
        const checkboxAll = checkboxes.first();

        checkboxAll.simulate('change', { target: { checked: true } });
        expect(getSelections(wrapper)).toEqual([0, 1, 2, 3]);

        checkboxes.at(1).simulate('change', { target: { checked: false } });
        expect(getSelections(wrapper)).toEqual([1, 2, 3]);

        checkboxes.at(1).simulate('change', { target: { checked: true } });
        expect(getSelections(wrapper)).toEqual([0, 1, 2, 3]);
    });

    it('fires change & select events', () => {
        const order = [];
        const handleChange = jest.fn().mockImplementation(() => {
            order.push('onChange');
        });
        const handleSelect = jest.fn().mockImplementation(() => {
            order.push('onSelect');
        });
        const rowSelection = {
            onChange: handleChange,
            onSelect: handleSelect,
        };
        const wrapper = mount(createTable({ rowSelection }));

        wrapper
            .find('input')
            .last()
            .simulate('change', { target: { checked: true } });

        expect(handleChange).toHaveBeenCalledWith(
            [3],
            [{ key: 3, name: 'Jerry' }]
        );
        expect(handleSelect.mock.calls.length).toBe(1);
        expect(handleSelect.mock.calls[0][0]).toEqual({
            key: 3,
            name: 'Jerry',
        });
        expect(handleSelect.mock.calls[0][1]).toEqual(true);
        expect(handleSelect.mock.calls[0][2]).toEqual([
            { key: 3, name: 'Jerry' },
        ]);
        expect(handleSelect.mock.calls[0][3].type).toBe('change');
        expect(order).toEqual(['onSelect', 'onChange']);
    });

    it('could hide selectAll checkbox and custom selection', () => {
        const rowSelection = {
            hideSelectAll: true,
        };
        const wrapper = mount(createTable({ rowSelection }));
        expect(wrapper.find('.selection').exists()).toBeFalsy();
    });

    it('handles disabled checkbox correctly when dataSource changes', () => {
        const rowSelection = {
            getCheckboxProps: (record) => ({ disabled: record.disabled }),
        };
        const wrapper = mount(createTable({ rowSelection }));
        const newData = [
            { key: 0, name: 'Jack', disabled: true },
            { key: 1, name: 'Lucy', disabled: true },
        ];
        wrapper.setProps({ dataSource: newData });
        wrapper.find('input').forEach((checkbox) => {
            expect(checkbox.props().disabled).toBe(true);
        });
    });

    it('highlight selected row', () => {
        const wrapper = mount(createTable());
        wrapper
            .find('input')
            .at(1)
            .simulate('change', { target: { checked: true } });
        expect(
            wrapper.find('tbody tr').at(0).hasClass('tableRowSelected')
        ).toBe(true);
    });

    it('fix selection column on the left', () => {
        const wrapper = mount(
            createTable({
                rowSelection: { fixed: true },
                scroll: { x: 903 },
            })
        );

        expect(wrapper.render()).toMatchSnapshot();
    });

    it('fix expand on th left when selection column fixed on the left', () => {
        const wrapper = mount(
            createTable({
                expandable: {
                    expandedRowRender() {
                        return <div />;
                    },
                },
                rowSelection: { fixed: true },
                scroll: { x: 903 },
            })
        );

        expect(wrapper.render()).toMatchSnapshot();
    });

    it('fix selection column on the left when any other column is fixed', () => {
        const wrapper = mount(
            createTable({
                rowSelection: {},
                columns: [
                    {
                        title: 'Name',
                        dataIndex: 'name',
                        fixed: 'left',
                    },
                ],
                scroll: { x: 903 },
            })
        );

        expect(wrapper.render()).toMatchSnapshot();
    });

    it('use column as selection column when key is `selection-column`', () => {
        const wrapper = mount(
            createTable({
                rowSelection: {},
                columns: [
                    {
                        title: 'Name',
                        dataIndex: 'name',
                        key: 'selection-column',
                    },
                ],
            })
        );

        expect(wrapper.render()).toMatchSnapshot();
    });

    it('should keep all checked state when remove item from dataSource', () => {
        const wrapper = mount(
            <Table
                rowSelection={{
                    selectedRowKeys: [0, 1, 2, 3],
                }}
                columns={columns}
                dataSource={data}
            />
        );
        expect(wrapper.find(CheckBox).length).toBe(5);
        wrapper.find(CheckBox).forEach((checkbox) => {
            expect(checkbox.props().checked).toBe(true);
        });
        wrapper.setProps({
            dataSource: data.slice(1),
            rowSelection: {
                selectedRowKeys: [1, 2, 3],
            },
        });
        expect(wrapper.find(CheckBox).length).toBe(4);
        wrapper.find(CheckBox).forEach((checkbox) => {
            expect(checkbox.props().checked).toBe(true);
        });
    });

    it('add columnTitle for rowSelection', () => {
        const wrapper = mount(
            <Table
                columns={columns}
                dataSource={data}
                rowSelection={{
                    columnTitle: '多选',
                }}
            />
        );
        expect(wrapper.find('thead tr th').at(0).text()).toBe('多选');
        wrapper.setProps({
            rowSelection: {
                type: 'radio',
                columnTitle: '单选',
            },
        });
        expect(wrapper.find('thead tr th').at(0).text()).toBe('单选');
    });

    it('select by checkbox to trigger stopPropagation', () => {
        const wrapper = mount(createTable());
        expect(() => {
            wrapper.find('span').at(10).simulate('click');
        }).not.toThrow();
    });

    describe('supports children', () => {
        const dataWithChildren = [
            { key: 0, name: 'Jack' },
            { key: 1, name: 'Lucy' },
            { key: 2, name: 'Tom' },
            {
                key: 3,
                name: 'Jerry',
                children: [
                    {
                        key: 4,
                        name: 'Jerry Jack',
                    },
                    {
                        key: 5,
                        name: 'Jerry Lucy',
                    },
                    {
                        key: 6,
                        name: 'Jerry Tom',
                        children: [
                            {
                                key: 7,
                                name: 'Jerry Tom Jack',
                            },
                            {
                                key: 8,
                                name: 'Jerry Tom Lucy',
                            },
                            {
                                key: 9,
                                name: 'Jerry Tom Tom',
                            },
                        ],
                    },
                ],
            },
        ];
        describe('supports checkStrictly', () => {
            it('should support `childrenColumnName`', () => {
                const onChange = jest.fn();

                const table = createTable({
                    dataSource: [
                        {
                            key: 0,
                            name: 'Jack',
                            childList: [
                                { key: 1, name: 'Light' },
                                { key: 2, name: 'Bamboo' },
                            ],
                        },
                    ],
                    expandable: {
                        childrenColumnName: 'childList',
                        defaultExpandAllRows: true,
                    },
                    rowSelection: {
                        checkStrictly: false,
                        onChange,
                    },
                });
                const wrapper = mount(table);
                const checkboxes = wrapper.find('input');
                expect(checkboxes).toHaveLength(1 + 3);

                checkboxes
                    .at(1)
                    .simulate('change', { target: { checked: true } });
                expect(getSelections(wrapper)).toEqual([0, 1, 2]);
            });
        });
    });

    describe('cache with selected keys', () => {
        it('default not cache', () => {
            const onChange = jest.fn();
            const wrapper = mount(
                <Table
                    dataSource={[{ name: 'light' }, { name: 'bamboo' }]}
                    rowSelection={{ onChange }}
                    rowKey="name"
                />
            );

            wrapper
                .find('tbody input')
                .first()
                .simulate('change', { target: { checked: true } });
            expect(onChange).toHaveBeenCalledWith(
                ['light'],
                [{ name: 'light' }]
            );

            wrapper.setProps({ dataSource: [{ name: 'bamboo' }] });
            wrapper
                .find('tbody input')
                .first()
                .simulate('change', { target: { checked: true } });
            expect(onChange).toHaveBeenCalledWith(
                ['bamboo'],
                [{ name: 'bamboo' }]
            );
        });

        it('cache with preserveSelectedRowKeys', () => {
            const onChange = jest.fn();
            const wrapper = mount(
                <Table
                    dataSource={[{ name: 'light' }, { name: 'bamboo' }]}
                    rowSelection={{ onChange, preserveSelectedRowKeys: true }}
                    rowKey="name"
                />
            );

            wrapper
                .find('tbody input')
                .first()
                .simulate('change', { target: { checked: true } });
            expect(onChange).toHaveBeenCalledWith(
                ['light'],
                [{ name: 'light' }]
            );

            wrapper.setProps({ dataSource: [{ name: 'bamboo' }] });
            wrapper
                .find('tbody input')
                .first()
                .simulate('change', { target: { checked: true } });
            expect(onChange).toHaveBeenCalledWith(
                ['light', 'bamboo'],
                [{ name: 'light' }, { name: 'bamboo' }]
            );
        });

        it('works with receive selectedRowKeys fron [] to undefined', () => {
            const onChange = jest.fn();
            const dataSource = [{ name: 'Jack' }];
            const wrapper = mount(
                <Table
                    dataSource={dataSource}
                    rowSelection={{ onChange, selectedRowKeys: [0] }}
                    rowKey="name"
                />
            );

            wrapper.setProps({
                rowSelection: { onChange, selectedRowKeys: undefined },
            });
            wrapper
                .find('tbody input')
                .first()
                .simulate('change', { target: { checked: true } });
            expect(onChange).toHaveBeenCalledWith(['Jack'], [{ name: 'Jack' }]);
        });

        it('works with selectionType radio receive selectedRowKeys from [] to undefined', () => {
            const onChange = jest.fn();
            const dataSource = [{ name: 'Jack' }];
            const wrapper = mount(
                <Table
                    dataSource={dataSource}
                    rowSelection={{
                        onChange,
                        selectedRowKeys: [0],
                        type: 'radio',
                    }}
                    rowKey="name"
                />
            );

            wrapper.setProps({
                rowSelection: {
                    onChange,
                    selectedRowKeys: undefined,
                    type: 'radio',
                },
            });
            wrapper
                .find('tbody input')
                .first()
                .simulate('change', { target: { checked: true } });
            expect(onChange).toHaveBeenCalledWith(['Jack'], [{ name: 'Jack' }]);
        });

        it('selectedRows and selectedKeys should keep sync in initial state', () => {
            const dataSource = [
                { name: 'Jack' },
                { name: 'Tom' },
                { name: 'Lucy' },
                { name: 'John' },
            ];
            const onChange = jest.fn();
            const rowSelection = {
                preserveSelectedRowKeys: true,
                onChange,
                selectedRowKeys: ['Jack'],
            };
            const wrapper = mount(
                <Table
                    dataSource={dataSource.slice(0, 2)}
                    rowSelection={rowSelection}
                    rowKey="name"
                    columns={[
                        {
                            title: 'Name',
                            dataIndex: 'name',
                            key: 'name',
                        },
                    ]}
                />
            );

            wrapper.setProps({
                dataSource: dataSource.slice(2, 4),
            });
            wrapper
                .find('tbody input')
                .first()
                .simulate('change', { target: { checked: true } });
            expect(onChange).toHaveBeenCalledWith(
                ['Jack', 'Lucy'],
                [{ name: 'Jack' }, { name: 'Lucy' }]
            );
        });
    });
});
