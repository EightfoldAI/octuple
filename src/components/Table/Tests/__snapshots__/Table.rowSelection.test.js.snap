version https://git-lfs.github.com/spec/v1
oid sha256:004cb87210a4e628e4926e9d6950549c68b4b82cc03f2aa76d47f24dd40320ae
size 25485810
